docs/src/book.md

```diff
@@ -0,0 +1,196 @@
+    1
+    4
+    0
+    1
+```
+
+We may now attempt to decompose this character type `h` by computing its scalar product with the irreducible character types.
+This returns a “generic” scalar product, plus a (possibly empty)
+list of parameter exceptions for which the general result may not hold.
+For example:
+```jldoctest book
+julia> scalar(T[4], h)
+0
+With exceptions:
+  2*n1 ∈ (q - 1)ℤ
+```
+This scalar product is $0$
+except when $q-1$ divides $2n_1$, where $n_1$ indicates the value of the parameter $n$ for the first factor in the scalar product (i.e. the character type $\chi_4$).
+For a “generic decomposition” we need to compute all the scalar products and
+exceptions.
+```jldoctest book
+julia> print_decomposition(T, h)
+Decomposing character chi:
+  <1,chi> = 1
+  <2,chi> = 2
+  <3,chi> = 2
+  <4,chi> = 0
+    With exceptions:
+      2*n1 ∈ (q - 1)ℤ
+  <5,chi> = 0
+    With exceptions:
+      2*n1 ∈ (q - 1)ℤ
+  <6,chi> = 0
+    With exceptions:
+      m1 + n1 ∈ (q - 1)ℤ
+      2*m1 - n1 ∈ (q - 1)ℤ
+      n1 ∈ (q - 1)ℤ
+      m1 ∈ (q - 1)ℤ
+      m1 - n1 ∈ (q - 1)ℤ
+      m1 - 2*n1 ∈ (q - 1)ℤ
+  <7,chi> = 0
+    With exceptions:
+      n1 ∈ (q - 1)ℤ
+  <8,chi> = 0
+    With exceptions:
+      (q + 1)*n1 ∈ (q^2 + q + 1)ℤ
+      q*n1 ∈ (q^2 + q + 1)ℤ
+      n1 ∈ (q^2 + q + 1)ℤ
+```
+
+This suggest a decomposition of $\chi_2\otimes\chi_2$ into $\chi_1+2\chi_2+2\chi_3$ “in general”.
+But comparing the respective degrees, we notice a discrepancy:
+```jldoctest book
+julia> degree(lincomb([1,2,2],[T[1],T[2],T[3]]))
+2*q^3 + 2*q^2 + 2*q + 1
+
+julia> degree(h)
+q^4 + 2*q^3 + q^2
+```
+
+To resolve this, we need to work through the exceptions.
+Recall that earlier we saw that $\langle\chi_4,\chi_2^2\rangle=0$
+except when $q-1$ divides $2n_1$, where $n_1$ was the value of the parameter $n$
+used in $\chi_4$. Further restrictions apply to $n$:
+```jldoctest book
+julia> printcharparam(T,4)
+4	n ∈ {1,…, q - 1} except n ∈ (q - 1)ℤ
+```
+
+So $n$ may take on any value between $1$ and $q-1$ not divisible by $q-1$.
+Hence the only possible exception is $n=(q-1)/2$ which can only
+occur if $q$ is odd.
+It thus makes sense to consider $q$ odd and $q$ even separately.
+
+We demonstrate this for $q$ even. Then $\langle\chi_4,\chi_2^2\rangle=0$
+and with a similar argument $\langle\chi_5,\chi_2^2\rangle=0$.
+We now construct a copy of the table but with the congruence equation $q\equiv 0\pmod 2$ applied:
+```jldoctest book
+julia> T2 = setcongruence(T, (0,2));
+
+```
+Inspecting the list of exceptions for $\langle\chi_6,\chi_2^2\rangle$, the first
+occurs when $q-1$ divides $m+n$. To study this case we specialize $m$:
+```jldoctest book
+julia> (q, (a, b, m, n)) = params(T2);
+
+julia> x = param(T2, "x");  # create an additional "free" variable
+
+julia> speccharparam!(T2, 6, m, -n + (q-1)*x)  # force m = -n (mod q-1)
+```
+
+Recomputing the scalar product gives a new result
+```jldoctest book
+julia> h = tensor(T2[2],T2[2]);
+
+julia> scalar(T2[6], h)
```

Comment:
This currently fails if I don't redefine `h` above because `h` belongs to the table `T`.

@SoongNoonien It would be nice to be able to write something like `T2(h)` to coerce `h` into the new table, *if the two table match up to the congruence.


Alternatively, perhaps it would be better to avoid the issue altogether by handling user specific congruences completely differently? I.e. instead of letting `setcongruence` produce a new table, and `speccharparam!` *modifying* a table, do it somehow else. But I am not sure how, and I think it depends a lot on practical use cases, so I'd be slow in changing this

Reply:
I'm not sure if this is any good, but I'd suggest to let `setcongruence` (or better `setcongruence!`) modify the table. (We haven't decided on a behavior yet, but I'd allow setting a congruence if and only if it implies the old one. So for example allow any congruence if it previously was `nothing`, allow `(0, 2)` if it previously was `(2, 4)` but disallow `(1, 6)` if it previously was `(1, 3)`.) And I'd rename `speccharparam!` to `speccharparam` and make it return the specialized character. But this doesn't solve the problem with `specclassparam!`. I don't know the use cases for `specclassparam!`, maybe it can even be removed?

Reply:
I agree in principle, though:
- we need to keep `specclassparam!`
- `speccharparam!` would be better called `specialize_params` or just `specialize` (the "character" bit would be implied by the type of the first argument, which would be a generic character)
- but for a user ideally there'd be a variant which takes a generic character, and a parameter exception (or a vector of parameter exceptions), and deals with it, so that the user doesn't have to come up with intermediate parameters like `x`

Reply:
> It would be nice to be able to write something like `T2(h)` to coerce h into the new table, *if the two table match up to the congruence.

This is possible now.